File: frontend/src/types/evaluation.ts
// 索引评测相关类型，命名与后端 schema 一致

export interface BenchmarkRequest {
  index_id: number;
  num_queries: number;
  top_k_list: number[];
  concurrency_list: number[];
}

export interface LatencyStats {
  concurrency: number;
  p50_ms: number;
  p95_ms: number;
  p99_ms: number;
  qps: number;
  mean_ms: number;
  total_queries: number;
}

export interface BenchmarkResult {
  index_id: number;
  dataset_id: number | null;
  backend: string;
  metric: string | null;
  build_time_seconds: number | null;
  memory_mb: number | null;
  num_queries: number;
  recalls: Record<string, number>;
  latencies: LatencyStats[];
  finished_at: string | null;
}

export interface BenchmarkTaskHandle {
  task_id: string;
  index_id: number;
  status: string;
}

export interface BenchmarkSummary {
  index_id: number;
  dataset_id: number | null;
  backend: string;
  recalls: Record<string, number>;
  finished_at: string | null;
}

// 检索日志统计相关类型，严格 snake_case 对齐后端 /api/v1/stats/search

export interface DatasetStat {
  dataset_id: number;
  dataset_name: string | null;
  total_queries: number;
  avg_latency_ms: number;
  p95_latency_ms: number;
}

export interface HourlyBucket {
  hour_iso: string;
  queries: number;
  avg_latency_ms: number;
}

export interface SearchStats {
  total_queries: number;
  overall_avg_latency_ms: number;
  overall_p95_latency_ms: number;
  by_dataset: DatasetStat[];
  hourly_24h: HourlyBucket[];
}

// v1.2 C3 · recall-QPS 帕累托扫描相关类型

export interface SweepRequest {
  dataset_id: number;
  backends: string[];
  top_k?: number;
  query_count?: number;
  ef_search_grid?: number[] | null;
  nprobe_grid?: number[] | null;
}

export interface SweepPoint {
  id: number;
  backend: string;
  params_json: Record<string, number | string | boolean>;
  recall: number;
  qps: number;
  p50_ms: number;
  p95_ms: number;
  p99_ms: number | null;
  mem_mb: number;
  on_pareto: boolean;
  created_at: string;
}

export interface SweepRun {
  id: number;
  dataset_id: number;
  created_by: number | null;
  status: 'pending' | 'running' | 'done' | 'failed';
  top_k: number;
  query_count: number;
  started_at: string;
  finished_at: string | null;
  error: string | null;
  created_at: string;
  points: SweepPoint[];
  pareto_count: number;
}

File: backend/app/schemas/evaluation.py
"""索引评测相关 schema。"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class BenchmarkRequest(BaseModel):
    """发起索引基准评测请求。"""

    index_id: int = Field(..., description="待评测索引 ID")
    num_queries: int = Field(100, ge=1, le=10000, description="评测使用的查询样本数量")
    top_k_list: list[int] = Field(
        default_factory=lambda: [10, 100],
        description="Recall 评测使用的 Top-K 取值列表",
    )
    concurrency_list: list[int] = Field(
        default_factory=lambda: [1, 4, 8, 16],
        description="并发压测使用的并发数列表",
    )


class LatencyStats(BaseModel):
    """单一并发档位的延迟统计。"""

    concurrency: int = Field(..., description="并发线程数")
    p50_ms: float = Field(..., description="P50 延迟（毫秒）")
    p95_ms: float = Field(..., description="P95 延迟（毫秒）")
    p99_ms: float = Field(..., description="P99 延迟（毫秒）")
    qps: float = Field(..., description="平均 QPS")
    mean_ms: float = Field(..., description="平均延迟（毫秒）")
    total_queries: int = Field(..., description="本档位总查询数")


class BenchmarkResult(BaseModel):
    """索引评测完整结果。"""

    index_id: int = Field(..., description="索引 ID")
    dataset_id: int | None = Field(None, description="对应数据集 ID")
    backend: str = Field(..., description="索引后端")
    metric: str | None = Field(None, description="距离度量")
    build_time_seconds: float | None = Field(None, description="索引构建耗时（秒）")
    memory_mb: float | None = Field(None, description="索引内存占用（MB）")
    num_queries: int = Field(..., description="实际评测的查询数")
    recalls: dict[str, float] = Field(
        default_factory=dict,
        description="不同 Top-K 下的 Recall，键为 K 的字符串形式",
    )
    latencies: list[LatencyStats] = Field(
        default_factory=list, description="按并发档位统计的延迟与 QPS"
    )
    finished_at: datetime | None = Field(None, description="评测完成时间")


class BenchmarkTaskHandle(BaseModel):
    """评测任务入队句柄。"""

    task_id: str = Field(..., description="ARQ 任务 ID，可用于结果查询")
    index_id: int = Field(..., description="本次评测的索引 ID")
    status: str = Field("queued", description="任务状态")


class SearchLogByDataset(BaseModel):
    """按数据集聚合的检索日志统计。"""

    dataset_id: int = Field(..., description="数据集 ID")
    dataset_name: str = Field(..., description="数据集名称，缺失时回退为 ``#ID``")
    total_queries: int = Field(..., description="该数据集下的检索次数")
    avg_latency_ms: float = Field(..., description="平均延迟（毫秒）")
    p95_latency_ms: float = Field(..., description="P95 延迟（毫秒）")


class SearchLogHourBucket(BaseModel):
    """最近 24 小时内单个小时桶的检索统计。"""

    hour_iso: str = Field(..., description="桶起点 ISO 时间戳（UTC, ``Z`` 后缀）")
    queries: int = Field(..., description="该小时桶内的检索次数")
    avg_latency_ms: float = Field(..., description="该桶平均延迟（毫秒），无数据为 0")


class SearchLogStats(BaseModel):
    """检索日志聚合统计响应。"""

    total_queries: int = Field(..., description="用户全部检索次数")
    overall_avg_latency_ms: float = Field(..., description="全部检索平均延迟（毫秒），无数据为 0")
    overall_p95_latency_ms: float = Field(..., description="P95 延迟（毫秒），无数据为 0")
    by_dataset: list[SearchLogByDataset] = Field(
        default_factory=list, description="按数据集分组的检索统计"
    )
    hourly_24h: list[SearchLogHourBucket] = Field(
        default_factory=list,
        description="最近 24h 滚动 1 小时窗口（共 24 桶，数组最后一个为当前桶）",
    )
